### PR TITLE
Milestone 3: Friendship v2

### PR DESCRIPTION
**Pull request by kcotrinam**
**In this milestone we:**
-  Styled the buttons and navbar.
- We created the pending requests counter.
- Add a button to accept the invitation on the user page.

### CODE REVIEW

**Review by kevcap — CHANGES_REQUESTED** [review]
# [STATUS] CHANGES REQUIRED♻️

#### Hey team,  your project is looking really good, however, to accomplish as the actual project, you required to do these changes, listed below:
 
## [REQUIRED CHANGES]
Please note that the user with 0 friends shouldn't be able to see any post. Currently when we create a new user, we're able to see posts made by users that aren't our friend yet.
- [x] PostsController#index displays posts in a timeline (their own posts + posts from their friends) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/timeline.png) from [Odin Project](https://www.theodinproject.com/courses/ruby-on-rails/lessons/final-project))

- [x] We have one rspec error, please fix it:
image

## Please make the required changes and resubmit for a new code review once you are ready💪

## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> gmail</a> [Twitter Follow](https://twitter.com/intent/follow?screen_name=kevinnnmn) [](https://microverse-students.slack.com/team/UT86D3AP4)

<sub>*We're here to help you, to ensure that you're using the best practices and getting you ready for the job market. In real-world, you'll have a lot of changes requests for your real clients, it's good to get used to it.</sub>

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hi @kcotrinam and @kenderb, great UI implementation

I am sorry the previous TSE missed out on some key implementations for **Friendship V2**, these implementations are what makes V2 differ from V1.

Should you face any challenge, please refer to this [article](https://explainextended.com/2009/03/07/selecting-friends) and [StackOverflow](https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table) discussion as a hint.

- [x] When friendship is confirmed, an additional row with reversed user-friend pair is created.
```ruby
def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

- [x] Thanks to this, we can find user friends only by user_id in friendships table
```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

- [x] User #pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

- [x] User #friend_requests can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

- [ ] In order to get all posts written by user on their friends, you can use simple method in User model (in order to NOT do it with Enumerables)
```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

#### That is it. I am sure you will learn a lot from it.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/views/layouts/_friend_request_number.html.erb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<div class="friends_request_numbers | has-background-danger has-text-white">
+  <p><%= current_user.inverted_friendships.length if current_user.inverted_friendships.length > 0%></p>
```
Comment (tirthajyoti-ghosh) [inline app/views/layouts/_friend_request_number.html.erb]: Please move this logic to helpers.

File: app/views/layouts/_navbar.html.erb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        <div class="buttons">
+          <a class="button is-primary">
+            <strong>Sign up</strong>
+          </a>
+          <a class="button is-light">
+            Log in
+          </a>
+        </div>
+      </div>
+      <% end %>
```
Comment (tirthajyoti-ghosh) [inline app/views/layouts/_navbar.html.erb]: There should be absolutely no logic in views. Please move them to helpers.

File: app/views/posts/_post.html.erb [hunk 3]
```diff
@@ -1,23 +1,30 @@
-<li class="post">
-  <div>
-    <div class='post-attribution' >
-      <%= link_to post.user.name, user_path(post.user), class: 'post-author' %>
-    </div>
-    <div class='post-liking'>
-      <span class='post-date'> <%= post.created_at.strftime("%Y/%m/%d") %> </span>
-      <%= like_or_dislike_btn(post) %>
+
+<%#if (current_user.friend?(post.user) || post.user.friend?(current_user)) || current_user == post.user%>
```
Comment (tirthajyoti-ghosh) [inline app/views/posts/_post.html.erb]: Please remove all commented codes.

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
# Changes required 🛑 

## Impressive progress so far 💪! Thanks for making all the changes 👏.

### However, we need to work on some changes before we can merge.

- The linters are not enabled in your project. Please go to the **Actions** tab and enable it. Also, fix any errors it displays.

- Please add tests for all the models (comment, like, post), not just the ones you created. Don't forget to add validation and association tests.

- Please add all the feature tests - accepting/rejecting friendship, commenting on posts, liking posts, creating posts, etc.

- There is some logic inside views. Please go through your views and move them to helpers.

Your project looks absolutely fantastic. Just make these changes and apply for another code review.

Happy coding 👨‍💻 

*Need help? Shoot me a message on Slack - [@Ghosh](https://app.slack.com/client/T47CT8XPG/DRDM6LAQN/user_profile/URB1MHXK6)*
<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Review by krys2fa — APPROVED** [review]
# Status: Project Approved! 🥇 
Congratulations on making all the requested changes and getting this milestone approved. 🏆

You can now merge this branch into the development branch and go on with your next project.

💡  Do not forget to submit your project completion form after merging this branch.

All the best team!

**Pull request merged**